### PR TITLE
Add edge based sub graph method.

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

New Behavior
----------------
Add edge based sub graph method to support link neighbor sampling.

### CODE REVIEW

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 1]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
             adj += np.eye(n)
 
         return unique_nodes, adj, src_nodes_idx
+
+
+def edge_sub_graph(graph: Graph, edges: np.ndarray, num_neighbors: List[int]):
+    """Sub graph of edges sampled from starting edges."""
+    edges = np.array(edges)
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/graph_ops.py:184]: `edges` is an np.ndarray according to type hint..

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 2]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
             adj += np.eye(n)
 
         return unique_nodes, adj, src_nodes_idx
+
+
+def edge_sub_graph(graph: Graph, edges: np.ndarray, num_neighbors: List[int]):
+    """Sub graph of edges sampled from starting edges."""
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/graph_ops.py:183]: could you describe argues, similar to `sample_out_edges` function above?

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 3]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
             adj += np.eye(n)
 
         return unique_nodes, adj, src_nodes_idx
+
+
+def edge_sub_graph(graph: Graph, edges: np.ndarray, num_neighbors: List[int]):
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/graph_ops.py:182]: Could you add a test, to show how to work with this function?

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 4]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
...
+    """Sub graph of edges sampled from starting edges."""
+    edges = np.array(edges)
+    subgraph = edges
+    for n in num_neighbors:
+        _, dst, types = edges[:, 0], edges[:, 1], edges[:, 2]
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/graph_ops.py:187]: Push this line outside of loop? I don't see how any of `dst`, `types`, `edges` is updated inside loop.

Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py:187]: Fixed, now edges is updated every loop.

File: src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import numpy.testing as npt
+from deepgnn.graph_engine.data.citation import Cora
+from deepgnn.graph_engine import graph_ops
+from deepgnn.graph_engine.snark.local import Client
+
+
+def test_edge_sub_graph():
+    random.seed(0)
+
+    # TODO tempfile
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py:18]: what tempfile?

File: src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    cora = Cora()
+    cl = Client(cora.data_dir(), [0])
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py:20]: Isn't `cora` already a graph? Why do you need to create another instance?

File: src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    edges = np.array([
+        [0, 1, 0],
+        [1, 2, 0],
+    ])
+
+    subgraph = graph_ops.edge_sub_graph(cl, edges, [5, 5])
+    print(subgraph)
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/graph_ops_test.py:28]: please, remove a stray print

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 8]
```diff
@@ -177,3 +177,27 @@ def sub_graph(
...
+        * graph: graph client.
+        * edges: np.ndarray of edges [[src, dst, type], ...].
+        * num_neighbors: list[int] Number of neighbors for each hop.
+    
+    Returns:
+        * np.ndarray of edges in subgraph [[src, dst, type], ...].
+    """
+    subgraph = edges
+    for n in num_neighbors:
+        _, dst, types = edges[:, 0], edges[:, 1], edges[:, 2]
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/graph_ops.py:196]: /s/`dst, types = edges[:, 1], edges[:, 2]`?

**Review by mortid0 — APPROVED** [review]
Please, update changelog.

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 9]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
...
+    """Sub graph of edges sampled from starting edges."""
+    edges = np.array(edges)
+    subgraph = edges
+    for n in num_neighbors:
+        _, dst, types = edges[:, 0], edges[:, 1], edges[:, 2]
+        dst_new = graph.sample_neighbors(dst, types, count=n)[0]
+        edges_new = np.stack((dst.repeat(n), dst_new.flatten(), types.repeat(n))).T
+        subgraph = np.concatenate((subgraph, edges_new))
+
+    subgraph = np.unique(subgraph, axis=0)
```
[the remark this answers is not in the document]
Comment (cozilla) [reply src/python/deepgnn/graph_engine/graph_ops.py:203]: why is the supgraph a list of edges? how will it be used for training?

File: src/python/deepgnn/graph_engine/graph_ops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py]: In new version of code, you are reviewing old code.

File: src/python/deepgnn/graph_engine/graph_ops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py]: Fixed in new code, this review is on old code.

File: src/python/deepgnn/graph_engine/graph_ops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py]: It is a ndarray of edges. Similar to sub_graph is a ndarray of nodes.

File: src/python/deepgnn/graph_engine/graph_ops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py:203]: This is what LinkNeighborLoader provides is the subgraph in the form of a list of edges.

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 10]
```diff
@@ -177,3 +177,18 @@ def sub_graph(
...
         return unique_nodes, adj, src_nodes_idx
+
+
+def edge_sub_graph(graph: Graph, edges: np.ndarray, num_neighbors: List[int]):
+    """Sub graph of edges sampled from starting edges."""
+    edges = np.array(edges)
+    subgraph = edges
+    for n in num_neighbors:
+        _, dst, types = edges[:, 0], edges[:, 1], edges[:, 2]
+        dst_new = graph.sample_neighbors(dst, types, count=n)[0]
```
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py:199]: Yes I can add a parmeter to change which version of sample_neighbors is being used.

File: src/python/deepgnn/graph_engine/graph_ops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (cozilla) [reply src/python/deepgnn/graph_engine/graph_ops.py:206]: the example shows it has a subgraph, not only a edge list.
```
   print(sampled_data)
        >>> Data(x=[1368, 1433], edge_index=[2, 3103], y=[1368],
                 train_mask=[1368], val_mask=[1368], test_mask=[1368],
                 edge_label_index=[2, 128], edge_label=[128])
```
https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/loader/link_neighbor_loader.html

[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py:206]: This is just the subgraph part not the whole dataloader

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 11]
```diff
@@ -177,3 +177,32 @@ def sub_graph(
...
+    Sub graph of edges sampled from starting edges.
+
+    Args:
+        * graph: graph client.
+        * edges: np.ndarray of edges [[src, dst, type], ...].
+        * num_neighbors: list[int] Number of neighbors for each hop.
+
+    Returns:
+        * np.ndarray of edges in subgraph [[src, dst, type], ...].
+    """
```
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/graph_ops.py:198]: strategy is the only one not verified, I think it is ok to skip since it is just passed to sample_neighbors and that is already tested.

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 12]
```diff
@@ -177,3 +177,33 @@ def sub_graph(
             adj += np.eye(n)
 
         return unique_nodes, adj, src_nodes_idx
+
+
+def edge_sub_graph(
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/graph_ops.py:182]: rename to edge list

File: src/python/deepgnn/graph_engine/graph_ops.py [hunk 13]
```diff
@@ -177,3 +177,33 @@ def sub_graph(
...
+    graph: Graph,
+    edges: np.ndarray,
+    num_neighbors: List[int],
+    strategy: str = "byweight",
+) -> np.ndarray:
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/graph_ops.py:187]: add type as parameter
